### PR TITLE
Generalize attribute syntax

### LINKED ISSUE #92: Parse error for some valid procedural attribute macros, derive macro helper attributes
`tree-sitter-rust` parses attributes using the [`meta-item`](https://doc.rust-lang.org/reference/attributes.html#meta-item-attribute-syntax) syntax, which is the syntax used by most of rust's built-in attributes. The `meta-item` syntax is a subset of the [`attr`](https://doc.rust-lang.org/reference/attributes.html#attributes) syntax. [Procedural attribute macros](https://doc.rust-lang.org/reference/procedural-macros.html#attribute-macros), however, use a larger subset of the `attr` syntax allowing arbitrary [`DelimTokenTree`s](https://doc.rust-lang.org/reference/macros.html). [Derive macro helper attributes](https://doc.rust-lang.org/reference/procedural-macros.html#derive-macro-helper-attributes) use the full `attr` syntax. As a result, `tree-sitter-rust` incorrectly parses many valid attributes.

As an example, the following is one of the usage examples for the [`thiserror`](https://crates.io/crates/thiserror) crate. `thiserror` defines a derive macro helper attribute, `error`:

```rust
#[derive(Error, Debug)]
pub enum Error {
    #[error("first letter must be lowercase but was {:?}", first_char(.0))]
    WrongCase(String),
    #[error("invalid index {idx}, expected at least {} and at most {}", .limits.lo, .limits.hi)]
    OutOfBounds { idx: usize, limits: Limits },
} 
```

The phrases `.limits.lo`, `.limits.hi`, and `first_char(.0)` are legal under the `attr` syntax but not under `meta-item`, and so these attributes are parsed incorrectly.

This is likely the cause of https://github.com/atom/atom/issues/20897. Any attribute not conforming to `meta-item` currently breaks `atom`'s rust syntax highlighting.

Attributes should instead be parsed using the more permissive `attr` syntax.

### PR DESCRIPTION
**Pull request by ninevra**
I attempted to fix #92. I didn't find a perfect solution; here's the resulting code, in case maintainers are interested.

I considered several approaches:
1. Parse attributes as [`meta-items`](https://doc.rust-lang.org/reference/attributes.html#meta-item-attribute-syntax) iff they match the `meta-item` syntax, and as general [`attrs`](https://doc.rust-lang.org/reference/attributes.html#attributes) otherwise. This preserves backwards compatibility, but creates a highly ambiguous grammar, since any `meta-item` is also a valid `attr`. I played around with `prec()`, `conflicts`, and `prec.dynamic()` but wasn't able to resolve the ambiguity correctly. Someone more familiar with LR/GLR parsing might be able to? In-progress code at https://github.com/ninevra/tree-sitter-rust/tree/ambiguitiy-and-precedence.
2. Parse [built-in attributes](https://doc.rust-lang.org/reference/attributes.html#built-in-attributes-index) as `meta-items`, and all others as `attrs`. This breaks backwards compatibility in that some custom attributes previously successfully parsed as `meta-items` and will now parse as `attrs`. It's arguably more correct than (1), since only built-in attributes are actually restricted to the `meta-item` syntax (see https://github.com/rust-lang/rust/pull/57367, https://github.com/rust-lang/rust/issues/57571). This is the approach taken in this PR.
3. Stop using the `meta-item` grammar at all and just use `attr`. This is fairly trivial, and I'd be happy to implement it if anyone wants. Breaks backwards compatibility for all attributes.

Are any of these options worth pursuing further?

### CODE REVIEW

**Comment by resolritter** [comment]
> Stop using the meta-item grammar at all and just use attr. This is fairly trivial, and I'd be happy to implement it if anyone wants. Breaks backwards compatibility for all attributes.

I think backwards compatibility is not a concern since Rust's grammar is still evolving and this project doesn't promise to be stable (still in the 0.x releases). For highlighting, selection and editing purposes, it shouldn't make a difference if those items are considered a meta-item or something else, as long as it makes sense. People's queries might need some slight update, but that's to be expected given the reasons I mentioned at the beginning.

> I didn't find a perfect solution

A few questions

- Can you go over why this isn't a "perfect solution"? What would be one?

I am thinking you implied this work is "imperfect" because you couldn't decide which way to go, but I would personally lean towards interpreting the word as referring to something "incomplete". Is that the case? Is the implementation not fully done for what you wanted to do, or what's the meaning of "imperfect" here?

- Which of the three approaches are you pursuing in this PR?

**Comment by ninevra** [comment]
> Can you go over why [option 3, changing the attribute grammar entirely] isn't a "perfect solution"? What would be one?

I was hoping for a solution that wouldn't change any existing successful parse results. If breaking changes are allowable, then option 3 seems fine to me. It does produce more consistent & accurate-to-spec results.

Its only other disadvantage is that the resulting parse (which consists of just nested tokentrees) isn't as useful to consumers as the meta-item grammar, which has more structure; consumers who _did_ make use of the meta-item grammar would have to re-parse the attr's arguments.

(A similar workflow is used in [`syn`](https://crates.io/crates/syn): [`Attributes`](https://docs.rs/syn/1.0.54/syn/struct.Attribute.html) are initially parsed with the general grammar, and then the attr's arguments can be re-parsed either as [`Meta`](https://docs.rs/syn/1.0.54/syn/enum.Meta.html) with [`Attribute::parse_meta`](https://docs.rs/syn/1.0.54/syn/struct.Attribute.html#method.parse_meta) or as a user-selected [`Parse`](https://docs.rs/syn/1.0.54/syn/parse/trait.Parse.html) type with [`Attribute::parse_args`](https://docs.rs/syn/1.0.54/syn/struct.Attribute.html#method.parse_args).) (I'm not familiar with `tree-sitter`'s end-user interface; is re-parsing a syntax tree node as some other type simple to do?)

**Comment by resolritter** [comment]
tree-sitter has Injections (https://tree-sitter.github.io/tree-sitter/syntax-highlighting#language-injection). One could "re-inject" the Rust parser inside of arbitrary nodes in order to parse them further. It's something I use, in practice, for highlighting interpolated code inside of JS strings. It's supported by the API already, but I don't know about how common of a practice this is.

Re-reading the OP

> 2. [...] It's arguably more correct than (1), since only built-in attributes are actually restricted to the meta-item syntax

I think 2 is a suitable solution as well, given this reasoning. If nothing's missing then you could move it out of Draft (AFAIK it's generally for unfinished work).

**Comment by ninevra** [comment]
Personally, I'd prefer option 3 if practical, since re-parsing attr arguments as the desired type is a cleaner interface than having to branch on whether the arguments are `meta-item` or token-tree.

Regardless, there are still design decisions to make: should the type of non-meta attr arguments be `token_tree`, or should it be some new non-terminal? (Currently, this PR uses `delim_token_tree` for the contents and `attr_item` for the wrapping type; I think it should be just `token_tree`, but I'd rather wait for input from a maintainer before putting more work into this.)

This PR is also currently blocked by #96, which introduces some of the same grammar productions (general token trees) and should be a much simpler/less controversial change.

**Comment by resolritter** [comment]
> I'm not familiar with `tree-sitter`'s end-user interface; is re-parsing a syntax tree node as some other type simple to do?)

@ninevra found this comment about reparsing on another issue

https://github.com/tree-sitter/tree-sitter-rust/issues/71#issuecomment-600172605

https://github.com/atom/atom/blob/0b34be796312a94331acbc4602944cfe3666c76a/packages/language-rust-bundled/lib/main.js#L2-L13

The injections are also showcased in the queries [here](https://github.com/tree-sitter/tree-sitter-rust/blob/40620bf4097cbc9cea79504d7e877865df43a19e/queries/injections.scm#L1). Yet another situation is what I've described in https://github.com/tree-sitter/tree-sitter-rust/pull/93#issuecomment-747863610.

**Comment by aryx** [comment]
@ninevra @resolritter this looks like a great PR. Would you mind rebasing on the latest to avoid the conflicts?
Also why is this still in a Draft state? This looks good to me.

**Comment by ninevra** [comment]
@aryx I've rebased this and updated it to reflect (afaict) the current spec, adding support for [`extended_key_value_attributes`](https://github.com/rust-lang/rust/issues/78835). Should be ready for review.

File: grammar.js
Comment (aryx) [inline grammar.js:38]: Would it be possible to add a link to a section of the rust reference manual that lists those builtin attributes?

**Review by aryx — APPROVED** [review]

File: grammar.js
Comment (ninevra) [reply grammar.js:38]: Sure? I'm not sure how best to do that given this PR has already been merged - should I create a second PR for that and any other changes that might be requested on review?

The reference section is https://doc.rust-lang.org/reference/attributes.html#built-in-attributes-index.

Keeping this updated is a little more complicated than checking that list; according to the current reference, only "most" built-in attributes follow the meta item syntax (although I haven't yet found one that doesn't), and additionally there may be other well-known attributes (e.g. tool attributes) that do use the meta item syntax.

Comment (aryx) [reply grammar.js:38]: You can make a second PR, or maybe because it's a small thing, you can fix it when you make another PR for fixing another problem.

**Pull request merged**
